File: Sources/KillerSudokuCore/PuzzleTimer.swift
import Foundation

/// A pausable stopwatch for the current puzzle. Pure value type — no wall-clock side effects of
/// its own — so callers (a SwiftUI `TimelineView`, say) decide when to re-read `elapsed()`
/// rather than this type pushing updates. `now` is injectable everywhere for deterministic tests.
public struct PuzzleTimer: Sendable {
    private var accumulated: TimeInterval = 0
    private var runningSince: Date?

    public init() {}

    public var isRunning: Bool { runningSince != nil }

    public mutating func start(now: Date = Date()) {
        guard runningSince == nil else { return }
        runningSince = now
    }

    public mutating func pause(now: Date = Date()) {
        guard let since = runningSince else { return }
        accumulated += now.timeIntervalSince(since)
        runningSince = nil
    }

    public mutating func reset() {
        accumulated = 0
        runningSince = nil
    }

    public func elapsed(now: Date = Date()) -> TimeInterval {
        guard let since = runningSince else { return accumulated }
        return accumulated + now.timeIntervalSince(since)
    }
}

File: desktop-kotlin/src/main/kotlin/me/gileskayo/killersudoku/core/PuzzleTimer.kt
package me.gileskayo.killersudoku.core

/** A pausable stopwatch for the current puzzle. Immutable value type -- no wall-clock side
 * effects of its own -- so callers decide when to re-read [elapsed] rather than this type
 * pushing updates. [nowEpochMillis] is injectable everywhere for deterministic tests. */
data class PuzzleTimer(
    private val accumulatedSeconds: Double = 0.0,
    private val runningSinceEpochMillis: Long? = null,
) {
    val isRunning: Boolean get() = runningSinceEpochMillis != null

    fun start(nowEpochMillis: Long = System.currentTimeMillis()): PuzzleTimer =
        if (runningSinceEpochMillis != null) this else copy(runningSinceEpochMillis = nowEpochMillis)

    fun pause(nowEpochMillis: Long = System.currentTimeMillis()): PuzzleTimer {
        val since = runningSinceEpochMillis ?: return this
        return copy(accumulatedSeconds = accumulatedSeconds + (nowEpochMillis - since) / 1000.0, runningSinceEpochMillis = null)
    }

    fun reset(): PuzzleTimer = PuzzleTimer()

    fun elapsed(nowEpochMillis: Long = System.currentTimeMillis()): Double {
        val since = runningSinceEpochMillis ?: return accumulatedSeconds
        return accumulatedSeconds + (nowEpochMillis - since) / 1000.0
    }
}
